### PR TITLE
Fatal Error if FS_METHOD is ftpext under PHP 8.0 or later for Admin l…

### PR DESCRIPTION
**Pull request by shoheitanaka**
Admin logging page fails with Fatal Error if FS_METHOD is ftpext under PHP 8.0 or later. This is because PHP 8.0 changed to send Fatal Error if null is given to ftp_nlist() which is called via WP_Filesystem under FS_METHOD is ftpext.

I wrote the corrected code based on the following correction pull request. https://github.com/tomusborne/generateblocks/pull/829

スクリーンショット 2024-02-22 4 09 41

### Submission Review Guidelines:
-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. I check the admin status loggin page, after I changed the FS_METHOD to ftpext and ssh2, direct for the server.
2. I check the admin status loggin page, after I changed the FS_METHOD to ftpsockets for another server.
3. I check the server error log, there were no errors.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Minor

#### Type

-   [ ] Fix - Fixed the admin logging page fails with Fatal Error if FS_METHOD is ftpext under PHP 8.0 or later.

#### Comment 

</details>

### CODE REVIEW

**Comment by shoheitanaka** [comment]
When I checked, it seems that a 'fatal error' occurs in all cases when using WC_Logging.
I think this is a high priority fix.

**Comment by coreymckrill** [comment]
@shoheitanaka Thanks for the PR! Interesting bug. I'm not sure the filesystem logger will work at all if `FS_METHOD` isn't set or defaulted to `direct`. We use some native PHP filesystem methods for some things when the WP Filesystem doesn't have an appropriate option, but it seems like these won't work if you're expecting to manage the filesystem via FTP.

So this makes me wonder:
* Should we automatically switch to the database logger if an alternative FS_METHOD is detected?
* Should we not use WP Filesystem methods at all, to ensure we're not mixing between direct and FTP?

In other words, I'm concerned that while your fix may resolve the immediate fatal error problem, it might not solve the larger issue. What do you think?

**Comment by shoheitanaka** [comment]
@coreymckrill I agree with your opinion.

Your first suggestion, which is to change to DB, is better.
However, the primary reason we do not recommend saving to the DB is how to think that WP cannot save it regarding connection errors with the DB.
It may be necessary to note that WP can save not all errors if you use it in the DB. If there is an error like this, even with text, WP will save not everything.

I suggest adding a note about logging.

**Comment by coreymckrill** [comment]
@shoheitanaka I don't have a quick and easy way to test using FTP for the filesystem. I'm curious if the logger will actually record entries to a file if the fix you have provided is in place and that fatal error doesn't happen. Do you have a way to test that?

**Comment by coreymckrill** [comment]
Resolved this issue with #45914 so I'm going to go ahead and close this one.
